### PR TITLE
Java SegmentTrie

### PR DESCRIPTION
**Pull request by ikitommi**
Rewrote the `reitit.segment` with Java. Tests say it's 2x faster now. Now only less than 3x slower than the [httprouter](https://github.com/julienschmidt/httprouter). Tested also against [calfpath](https://github.com/kumarshantanu/calfpath).

* added also `:inject-router?` and `:inject-match?` options to `reitit.ring/ring-router`, `reitit.http/http-router`, `reitit.http/routing-interceptor` => can be disabled for better perf (e.g. benchmarks).

New opensensors values:

opensensors

### CODE REVIEW

File: modules/reitit-ring/src/reitit/ring.cljc [hunk 1]
```diff
@@ -226,20 +226,54 @@
...
   and options map, with the following keys:
 
-  | key           | description |
-  | --------------|-------------|
-  | `:middleware` | Optional sequence of middleware that wrap the ring-handler"
+  | key               | description |
+  | ------------------|-------------|
+  | `:middleware`     | Optional sequence of middleware that wrap the ring-handler
+  | `:inject-match?`  | Boolean to inject `match` into request under `:reitit.core/match` key (default true)
+  | `:inject-router?` | Boolean to inject `router` into request under `:reitit.core/router` key (default true)"
```
[the remark this answers is not in the document]
Comment (ikitommi) [reply modules/reitit-ring/src/reitit/ring.cljc]: I described it in the changelog: perf. Those 2 injects take ~60ns on my laptop, which in worst case doubles the time to make a routing decision.

[the remark this answers is not in the document]
Comment (miikka) [reply modules/reitit-ring/src/reitit/ring.cljc]: Right

File: modules/reitit-core/java-src/reitit/SegmentTrie.java [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        pointer = s;
+      } else if (p.startsWith("*")) {
+        Keyword k = Keyword.intern(p.substring(1));
+        SegmentTrie s = pointer.catchAll.get(k);
+        if (s == null) {
+          s = new SegmentTrie();
+          s.data = data;
+          pointer.catchAll.put(k, s);
+        }
+        break;
```
[the remark this answers is not in the document]
Comment (ikitommi) [reply modules/reitit-core/java-src/reitit/SegmentTrie.java]: simplified, as there can only one catchAll per segment.

File: modules/reitit-core/src/reitit/segment.cljc [hunk 3]
```diff
@@ -43,14 +44,20 @@
                (if (and wilds? (not (str/blank? p))) (some #(-lookup (impl/fast-get children' %) ps (assoc path-params % p)) wilds))
                (if catch-all (-catch-all children' catch-all path-params p ps)))))))))
 
+;;
+;; public api
+;;
```
[the remark this answers is not in the document]
Comment (ikitommi) [reply modules/reitit-core/src/reitit/segment.cljc]: true, did.

File: modules/reitit-core/java-src/reitit/SegmentTrie.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (miikka) [reply modules/reitit-core/java-src/reitit/SegmentTrie.java]:
```clj
(-> (s/insert nil "/foo" {:a 1}) (s/compile) (s/lookup "/foo")) ; => #reitit.segment.Match{:data {:a 1}, :path-params {}}
(-> (s/insert nil "/foo" {:a 1}) (s/insert "/foo/*" {:b 1}) (s/compile) (s/lookup "/foo")) ; => nil
(-> (s/insert nil "/foo" {:a 1}) (s/insert "/foo/*" {:b 1}) (s/compile) (s/lookup "/foo/bar")) ; => #reitit.segment.Match{:data {:b 1}, :path-params {: "bar"}}
```

Is the `nil` result for the second expression really right? The first and third result look correct to me.

```clj
(def router (r/router [["/foo/:bar" {:a 1}] ["/foo/:bar/*" {:b 1}]]))
(r/match-by-path router "/foo/bar/baz")
; => #reitit.core.Match{:template "/foo/:bar/*", :data {:b 1}, :result nil, :path-params {:bar "bar", : "baz"}, :path "/foo/bar/baz"}
(r/match-by-path router "/foo/bar")
; => nil
```

**Comment by ikitommi** [comment]
Now I get it, fixed, and added tests. The created matcher for catchAll needed to be wrapped into `LinearMatcher` with data as first check.

**Comment by ikitommi** [comment]
https://github.com/metosin/reitit/pull/203/commits/b8d82862652356c8970193ca1aab01c51b122854

**Pull request merged**